coffee_maker_navigator_2/lib/di/di/dependency_container_manager.dart

```diff
@@ -0,0 +1,137 @@
+  /// [DependencyContainer] for the specified type [T].
+  final _containerFactories =
+      <Type, DependencyContainer Function(AppLevelDependencyContainer)>{};
+
+  // A map that tracks the subscribers for each container type.
+  final _containerSubscribers = <Type, Set<Object>>{};
+
+  // A map that holds the currently active containers by their type.
+  // An active container is one that is currently in use, meaning it has at least one subscriber.
+  final _activeContainers = <Type, DependencyContainer>{};
+
+  /// Returns the single instance of [DependencyContainerManager].
+  static DependencyContainerManager get instance => _instance;
+
+  /// Factory constructor to return the singleton instance.
+  factory DependencyContainerManager() => _instance;
+
+  /// Internal constructor for singleton implementation.
+  DependencyContainerManager._internal();
+
+  /// Initializes the app-level dependencies.
+  Future<void> init() async {
+    await _appLevelDependencies.init();
+    registerContainerFactory<AppLevelDependencyContainer>(
+        (_) => _appLevelDependencies);
+    subscribeToContainer<AppLevelDependencyContainer>(this);
+  }
+
+  /// Registers a dependency container factory for a specific type [T].
+  ///
+  /// This method allows you to register a factory function that will be responsible
+  /// for creating instances of the dependency container for the specified dependency container
+  /// type [T].
+  ///
+  /// [factory]: A function that takes an existing [DependencyContainer] and returns
+  /// a new instance of a [DependencyContainer] for the specified type [T].
+  void registerContainerFactory<T>(
+      DependencyContainer Function(DependencyContainer) factory) {
+    _containerFactories[T] = factory;
+  }
+
+  /// Adds a subscriber for the container of type [C] and manages the container's lifecycle.
+  ///
+  /// This method registers a subscriber for the specified container type [C] and
+  /// manages the lifecycle of the container based on the presence of subscribers.
+  ///
+  /// [subscriber]: An object that subscribes to the container of type [C]. The presence
+  /// of subscribers influences the creation and destruction of the container.
+  void subscribeToContainer<C>(Object subscriber) {
+    final currentSubscribers = _containerSubscribers[C] ??= {};
+    // Add the subscriber to the set of subscribers for the container type C if not already present.
+    if (currentSubscribers.add(subscriber)) {
+      _manageContainerLifecycle<C>();
+    }
+  }
+
+  /// Removes a subscriber for the container of type [C] and manages the container's lifecycle.
+  ///
+  /// This method unregisters a subscriber for the specified container type [C] and
+  /// manages the lifecycle of the container based on the remaining subscribers.
+  ///
+  /// [subscriber]: An object that unsubscribes from the container of type [C]. The absence
+  /// of subscribers can lead to the destruction of the container.
```

Comment:
I think `can` is not fit here, it is guaranteed

Reply:
done 5bcc8fd28fe071d7152e9d1f56a188c3a514de34